File: JavaVersion/CaesarCipher.java
public class CaesarCipher {
	
	// Stores the final encoded String. Use string builder as string
	// Concatenation is a more expensive operation
	
	private StringBuilder encoded = new StringBuilder();

	public static void main(String[] args) {

		CaesarCipher caesarCipher = new CaesarCipher();
		caesarCipher.run();
	}

	public void run() {
		
		System.out.println(new CaesarCipher().cipher("abc def", 3));
		System.out.println(new CaesarCipher().cipher(
				"THE QUICK BROWN FOX JUMPS OVER THE LAZY DOG", 3));

	}
	
	/*
	 * This method takes in a string and a shift value that adds
	 * the shift value to the original string and produces a new
	 * encoded string. 
	 * 
	 * E.g: - cipher("abc",3) = "def"
	 *  
	 */

	public String cipher(String str, int shift) {
		
		// Throw exception for null values and empty string
		if (str == null || str == "") {
			throw new IllegalArgumentException("Invalid Input");
		}
		
		
		for (int i = 0; i < str.length(); i++) {
			
			// encode the character by adding the shift value to it
			char c = (char) (str.charAt(i) + shift);
			
			// Check if it is an uppercase character or Lowercase character
			// If its not either of them, then the character is not a letter and
			// and hence we just append them. My assumption is that if the
			// character is not a letter, then we don't encode them and just append
			// the original character
			
			if (Character.isUpperCase(str.charAt(i))) {
				encodeCharacter(c, 'Z');
			} else if (Character.isLowerCase(str.charAt(i))) {
				encodeCharacter(c, 'z');
			} else {
				encoded.append((char) str.charAt(i));
			}

		}
		return encoded.toString();
	}
	
	/*
	 * This function takes in an encoded character and the last character in
	 * the english alphabet. It compares both of them to determine if there
	 * is an overflow of a character. For Example, if originalcharacter is x and
	 * we shift it by 3 times, the encoded character should be a. so If x + 3 > z,
	 * we subtract 26 from it to get the wrapped around character. 
	 * 
	 */

	private void encodeCharacter(Character encodedCharacter,
			Character lastCharacter) {
		
		if (encodedCharacter > lastCharacter) {
			encoded.append((char) (encodedCharacter - 26));
		} else {
			encoded.append((char) (encodedCharacter));
		}
	}

}

File: RubyVersion/CaesarCipher.rb
class CaesarCipher
	
	def initialize(encoded)
		@encoded = encoded 	#instance variable stores the encoded result
	end
	
	
	#This method takes in a string and a shift value that adds
	#the shift value to the original string and produces a new
	#encoded string.   
	#E.g: - cipher("abc",3) = "def"
	 
	def cipher(str,shift)
	
		if str == '' or str.nil? or !str.is_a? String
			raise "Invalid Input"
		end 
		
		str.each_byte do |c|
		
			original_character = c.chr
			
			# encode the character by adding the shift value to it
			encoded_character = (c + shift).chr
			
			# Check if it is an uppercase character or Lowercase character
			# If its not either of them, then the character is not a letter and
			# and hence we just append them. My assumption is that if the
			# character is not a letter, then we don't encode them and just append
			# the original character
			
			if is_upper?(original_character)
				encodeCharacter(encoded_character,'Z')
			elsif is_lower?(original_character)
				encodeCharacter(encoded_character.chr,'z')
			else
				# use concat to conacatenate strings as it is much faster than
				# the + operator for conacatenation
				@encoded.concat(original_character)
			end
			
		end
		
		return @encoded
	end
	
	private
		
		# This function takes in an encoded character and the last character in
		# the english alphabet. It compares both of them to determine if there
		# is an overflow of a character. For Example, if originalcharacter is x and
		# we shift it by 3 times, the encoded character should be a. so If x + 3 > z,
		# we subtract 26 from it to get the wrapped around character. 
			
		def encodeCharacter(encoded_character, last_character)
			if (encoded_character > last_character)
				@encoded.concat((encoded_character.ord - 26).chr)
			else	
				@encoded.concat(encoded_character)
			end
		end

	private
		# determines if the character is lowercase.
		# Note that this function is needed as space is 
		# considered as an uppercase as well as a lowercase character
		# in ruby .upcase or .downcase api
		
		def is_lower?(c)
			return c >= 'a' && c <= 'z'
		end
			
	private
		
		# determines if the character is uppercase.
		# Note that this function is needed as space is 
		# considered as an uppercase as well as a lowercase character
		# in ruby if we use .upcase or .downcase api
		
		def is_upper?(c)
			return c >= 'A' && c <= 'Z'
		end
		
end


def main
	puts CaesarCipher.new('').cipher('xyz',3)
	puts CaesarCipher.new('').cipher('abc def',3)
end

main